alembic/versions/43882376fe16_add_ranch_to_the_tf_groups.py

```diff
@@ -0,0 +1,142 @@
+from sqlalchemy.orm import (
+    relationship,
+)
+
+from alembic import op
+
+# https://github.com/python/mypy/issues/2477#issuecomment-313984522 ^_^
+if TYPE_CHECKING:
+    Base = object
+else:
+    Base = declarative_base()
+
+# revision identifiers, used by Alembic.
+revision = "43882376fe16"
+down_revision = "48b09a73cf77"
+branch_labels = None
+depends_on = None
+
+
+class TestingFarmResult(str, enum.Enum):
+    __test__ = False
+
+    new = "new"
+    queued = "queued"
+    running = "running"
+    passed = "passed"
+    failed = "failed"
+    skipped = "skipped"
+    error = "error"
+    unknown = "unknown"
+    needs_inspection = "needs_inspection"
+    retry = "retry"
+    complete = "complete"
+    canceled = "canceled"
+    cancel_requested = "cancel-requested"
+
+    @classmethod
+    def from_string(cls, value):
+        try:
+            return cls(value)
+        except ValueError:
+            return cls.unknown
+
+
+class TFTTestRunGroupModel(Base):
+    __tablename__ = "tft_test_run_groups"
+    id = Column(Integer, primary_key=True)
+    submitted_time = Column(DateTime, default=datetime.utcnow)
+    ranch = Column(String)
+
+    # runs = relationship("pipelines", back_populates="test_run_group")
+    tft_test_run_targets = relationship(
+        "TFTTestRunTargetModel",
+        back_populates="group_of_targets",
+    )
+
+    @property
+    def grouped_targets(self) -> list["TFTTestRunTargetModel"]:
+        return self.tft_test_run_targets
+
+
+class TFTTestRunTargetModel(Base):
+    __tablename__ = "tft_test_run_targets"
+    id = Column(Integer, primary_key=True)
+    pipeline_id = Column(String, index=True)
+    identifier = Column(String)
+    status = Column(Enum(TestingFarmResult))
+    target = Column(String)
+    web_url = Column(String)
+    # datetime.utcnow instead of datetime.utcnow() because its an argument to the function
+    # so it will run when the model is initiated, not when the table is made
+    submitted_time = Column(DateTime, default=datetime.utcnow)
+    data = Column(JSON)
+    tft_test_run_group_id = Column(Integer, ForeignKey("tft_test_run_groups.id"), index=True)
+
+    # copr_builds = relationship(
+    #     "copr_build_targets",
+    #     secondary="tf_copr_build_association_table",
+    #     backref="tft_test_run_targets",
+    # )
+    # koji_builds = relationship(
+    #     "koji_build_targets",
+    #     secondary="tf_koji_build_association_table",
+    #     backref="tft_test_run_targets",
+    # )
+    group_of_targets = relationship(
+        "TFTTestRunGroupModel",
+        back_populates="tft_test_run_targets",
+    )
+
+
+def upgrade():
+    bind = op.get_bind()
+    session = orm.Session(bind=bind)
+
+    op.add_column("tft_test_run_groups", sa.Column("ranch", sa.String(), nullable=True))
+
+    # Populate the ranches
+    targets = select(TFTTestRunTargetModel)
+    for (target,) in session.execute(targets):
+        if target.web_url is None:
```

Comment:
just a nitpick, since we are settting this on the group level, we could skip if the group has the ranch already set?